### ISSUE TITLE
[Dropdown Menu] Incorrect use of aria menu roles

### ISSUE THREAD

**Issue by jimmynotjim** [zurb/foundation-sites#11453]
The aria menu roles are intended for application menus, not navigation menus, and are problematic with Apple's Voice Over features. WebAim provides this guidance:

> "If the navigation links at the top of a webpage have incorrectly been assigned an ARIA role='menu' (which is appropriate for application menus, not for navigation menus), this triggers forms mode, and the screen reader user cannot read with the arrow keys. The same thing can occur with other ARIA roles such as tab, dialog, grid, and toolbar."

_Source_ - https://webaim.org/blog/three-things-voiceover/ (__ARIA, screen reader modes, and VoiceOver__)

Overall, menu roles should be avoided for most websites, due to the nature of standards based web navigation.

## Expected Behavior

When navigating the dropdown menu using Voice Over's Start reading feature the navigation items should only be read aloud once.

## Current Behavior

When navigating the dropdown menu using Voice Over's Start reading feature the navigation items are read twice due to the use of `role=menuitem`.

## Possible Solution

1. Remove the `menubar`, `menuitem`, and `menu` roles.
1. Update the documentation to show navigation menus should always be wrapped in a `nav` element.

## Test Case and/or Steps to Reproduce (for bugs)

How to reproduce:
1. Navigate to https://codepen.io/jimmynotjim/full/VBNGbY/
1. Turn on voice over (cmd + f5)
1. Click into the main content to avoid the Codepen chrome
1. Turn on start reading feature (ctrl + opt + A)
1. Voice Over reads the text content of each link twice, once as the menu item, and again as a link
  <img width="595" alt="screen shot 2018-08-14 at 10 44 52 am" src="https://user-images.githubusercontent.com/1280430/44102441-4ae0dde0-9faf-11e8-9dbb-261e7c6ae0ac.png">
  <img width="597" alt="screen shot 2018-08-14 at 10 45 07 am" src="https://user-images.githubusercontent.com/1280430/44102448-4df8d992-9faf-11e8-8cfc-f68641ffc861.png">

## Context

I discovered the aria menu roles while working on GoFundMe's navigation which relies on Foundation's dropdown menu.

## Your Environment

- Foundation version(s) used: Foundation 6
- Browser(s) name and version(s): Any browser on Mac
- Operating System and version (desktop or mobile): macOS 10 (High Sierra) but I have observed this issue on every recent version of OSX and macOS.
- Link to your project: http://www.gofundme.com

## Checklist (all required):

- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] This is a bug report or a feature request.
- [x] There are no other issues similar to this one.
- [x] The issue title is descriptive.
- [x] The template is fully and correctly filled.

**Comment by ncoden** [comment]
Related to the discussions in https://github.com/zurb/foundation-sites/pull/11440

**Comment by jimmynotjim** [comment]
This might be a good reference for both this issue and #11440 

http://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html

Aria roles should be limited to fixing markup mistakes rather than being used by default. Since Foundation seems to keep to semantic markup, I don't really see a valid reason for relying on them in most situations.

**Comment by ncoden** [comment]
@jimmynotjim

I read this article and I confirm you this is a good reference. You can notice I referenced it in #11440.

IMHO most aria roles provide additional intormations to assistive tools like related elements and proposed keyboard interactions. @mcking65 words are quite helpful: 

> When choosing the pattern to use, one of the most important considerations is the keyboard interaction model. If you are not implementing a menu-like or tree-like keyboard interaction model, then do not use those roles and their associated patterns as described in the ARIA Authoring Practices. If you are implementing one of those keyboard interaction models, then please use the appropriate pattern so that screen reader users will be aware of the interactions that are supported. Also, please provide visual design that helps sighted users be aware of the keyboard interactions that are available.

I'll try to resolve #11440 first then will take a look at this issue after. Depending on what will be decided, we may have (or not) to rework on all the Foundation menu components.

**Comment by jimmynotjim** [comment]
@ncoden Yep, if your default markup creates a good keyboard experience, then aria roles aren't really necessary, but if there are issues that are impossible to work around, for whatever reason, that's when they start to come into importance very quickly. 

Take this example, it uses the Foundation grid for the layout but a custom menu I built to work through some of the issues we're seeing at GoFundMe using regular standard markup for as much as possible (it's only built for large screens at the moment).

https://codepen.io/jimmynotjim/full/qyGaWL/

You'll notice that I changed the trigger interaction to click rather than hover. In my experience determining the user's intent is most important and hover can be very ambiguous, plus it can be difficult for users who have dexterity issues and can't easily keep a mouse over the target. If you test with Voice Over you'll find it relies on the regular enter + tab interactions rather than the custom arrow up/down + escape interactions provided by `role=menu`.

If you're looking for help, I may be able to contribute in some off hours.

**Comment by jimmynotjim** [comment]
I read this some time ago, but just re-read it to fill in some gaps, it's got a lot of helpful info to understand the correct use of aria menus as well as options for navigation menus https://inclusive-components.design/menus-menu-buttons/

**Comment by ncoden** [comment]
@jimmynotjim I read it too, and referenced it in #11440. It was also published on Smashing Maganize there: https://www.smashingmagazine.com/2017/11/building-accessible-menu-systems.

Take a look at the "The menu and menuitem roles" section:

> ```js
> <ul role="menu">
>   <li role="presentation">
>     <a href="/" role="menuitem">Home</a>
>   </li>
> </ul>
> ```
> What we’re left with is a kind of hybrid component, which isn’t quite a true menu but at least tells users whether the list of links is open, thanks to the aria-expanded state. This is a perfectly satisfactory pattern for navigation menus.

The following menu example from the W3C uses the same pattern (except for the `presentation` role). Take a look: https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html.

However, I feels like repeating myself there. I will not reply on any further messages on this issue until #11440 is resolved. You are welcome to join the conversation. 😄

**Comment by JeremyEnglert** [comment]
@mcking65 is spot on. But in the case of Foundation, I think `role=navigation` makes much more sense. Also, the dropdown menus most likely need `aria-expanded` attributes added.

**Comment by mcking65** [comment]
The ARIA Authoring Practices Guide now includes an [Example Disclosure for Navigation Menus](http://w3c.github.io/aria-practices/examples/disclosure/disclosure-navigation.html).

This is an alternative pattern for site navigation bars. Comparing it to the menubar pattern:

1. Like menubar, it can include support for arrow keys among links.
2. Like menubar, escape will close dropdowns.
3. Unlike menubar, it is not a single tab stop; instead every element, when visible,  is Tab focusable.
4. Unlike menubar, it does not support first-letter navigation, e.g., type "c" to move focus to "Campus Tours" in the "About" dropdown. 
5. Unlike menubar, Because it uses really basic semantics that have been supported in browsers by screen readers for a long time, the disclosure pattern works correctly with any screen reader.

The reason it is not a single Tab stop and does not support first-letter navigation is because the dropdown is a list of links, not a widget designed to intercept keyboard input from users of screen readers that have an interaction mode, e.g., JAWS forms mode.

**Comment by JeremyEnglert** [comment]
Matt,

First off - thank you for all of your help (and for everything else you do)! I'm learning a ton from your posts.

In the "Example Disclosure for Navigation Menu" you posted, it appears a `<button>` is used for the top-level dropdown - is that a requirement? 

If so, how would you treat top level items that _do not_ have a dropdown?

**Comment by mcking65** [comment]
> Matt,
> 
> First off - thank you for all of your help (and for everything else you do)! I'm learning a ton from your posts.

Thank you, and you're welcome!!

> In the "Example Disclosure for Navigation Menu" you posted, it appears a `<button>` is used for the top-level dropdown - is that a requirement?

Yes, a button with aria-expanded set to reflect the state of the dropdown is the core feature of the pattern. You could use a link, div, or span  with role button if that is necessary.
 
> If so, how would you treat top level items that _do not_ have a dropdown?

They would be ordinary links with no additional ARIA.

**Comment by charmarkk** [comment]
So glad to have run into you over here @mcking65 - I was _literally just_ emailing with our designer about this being a problem and how to fix it (or file a bug). I'll see myself out since you've got this under control. ;)

**Comment by DanielRuf** [comment]
Hi @mcking65,

Do you want to provide a PR so we can fix this?

**Comment by mcking65** [comment]
> Hi @mcking65,
> 
> Do you want to provide a PR so we can fix this?


Sorry, hands over full over at aria-practices.

**Comment by JeremyEnglert** [comment]
I think the current menu is _technically_ correct. But lets keep all of this in mind when reworking the menus for V7.

**Comment by charmarkk** [comment]
I'm a little concerned and disappointed at this being closed, as it seems like the fix should be just moving a couple things around in the JS (but then I'm not a super skilled programmer). I'm not sure *I'd* call it technically correct, as the ARIA isn't correct, and I have a specific use case where it is *actively* interfering with a keyboard's and/or screen reader's ability to use the menu appropriately.

**Comment by philwolstenholme** [comment]
A client's site of ours has been audited under WCAG 2.1 and the auditors pointed out this issue.

> The roles "menubar" and "menuitem" should be reserved for navigation components in widgets. Remove these, and the "role="none" attributes from list items, so that navigation elements are links in an unordered list. The controls are not announced as links by screen reader software. Experienced and confident users will assume from context that they are links - but may not think they are “working” because they aren’t announced as links. Inexperienced or unconfident users may be a bit more confused.
